AAP-6908 update and refactor docs (#277)
ansible/ansible-rulebook: docs/conditions.rst

Conditions
==========

In event driven automation a condition determines if a rule fires (runs its action).

Example:

    .. code-block:: yaml

        condition: event.status == "enabled"

> Each of the condition(s) can use information from
> 
>  * Event received

Reviewer: ```suggestion
Each of the condition(s) can use information from
 * Event received
```

    .. code-block:: yaml

        name: One condition combining attributes
        condition: event.target_os == "linux" or event.target_os == "windows"
        action:
          debug:

Multiple conditions with facts and events and **all** of one of them have to match
----------------------------------------------------------------------------------

    .. code-block:: yaml

        name: Condition using both a fact and an event
        condition:
          all:
            - fact.meta.hosts == "localhost"
            - event.target_os == "windows"
        action:
          debug:

Condition with fact and event, fact being passed in via --variables on command line
-----------------------------------------------------------------------------------

    .. code-block:: yaml

        name: Condition using a passed in variable and an event
        condition:
          all:
            - facts.first << fact.custom.expected_index is defined
            - event.i == facts.first.custom.expected_index
        action:
          debug:

In the above example the custom.expected_index was passed in a variables file via ``--variables`` from the
command line to ansible-rulebook.

> | When evaluating a single event you can compare multiple
> | properties/attributes from the event using **and** or **or**

Reviewer: ```suggestion
| When evaluating a single event you can compare multiple properties/attributes from the event using **and** or **or**.
```

Author: suggestion rejected, many things are in the same way, out of the scope of this PR

Logical and
-----------
    .. code-block:: yaml

        name: Multiple Attribute match from a single event
        condition: event.target_os == "linux" and event.version == "1.1"
        action:
          debug:

Logical or
----------

    .. code-block:: yaml

        name: Match any one attribute from a single event
        condition: event.version == "2.0" or event.version == "1.1"
        action:
          debug:

| The "and" and "or" keywords are case sensitive. You can't use
| "AND" or "OR" for the logical operators.

Combining logical operators
----------

> You can combine multiple ``and`` operators

Reviewer: ```suggestion
You can combine multiple ``and`` operators:
```

    .. code-block:: yaml

        name: Combining and operators
        condition: event.version == "2.0" and event.name == "example" and event.alert_count > 10
        action:
          debug:

> If you combine ``and`` and ``or`` operators they must be enclosed in parenthesis

Reviewer: ```suggestion
If you combine ``and`` and ``or`` operators they must be enclosed in parenthesis:
```

    .. code-block:: yaml

        name: Combining and -and- or operators
        condition: ((event.i > 100 and event.i < 200) or (event.i > 500 and event.i < 600))
        action:
          debug:

    .. code-block:: yaml

        name: Combining and -and- or operators
        condition: (event.i > 100 and event.i < 200) or event.i > 1000
        action:
          debug:

Multiple conditions with assignment
-----------------------------------

When a condition is evaluated if the condition passes the matching event
it is stored in well known attribute(s) called **m**, **m_1**, **m_2**.....

> You can optionally alias these attribute(s) using the **<<** operator e.g

Reviewer: ```suggestion
You can optionally alias these attribute(s) using the **<<** operator. For example:
```

| outside of the rule. They can be used in assignments and accessing the saved
| values in a condition or in action.

Multiple condition with default assignments
-------------------------------------------

    .. code-block:: yaml

        name: multiple conditions
        condition:
          all:
            - event.i == 1
            - event.i == 2
            - event.i == events.m.i + 3
        action:
          debug:
            first: "{{events.m}}"
            second: "{{events.m_1}}"
            third: "{{events.m_2}}"

The first match is stored as **m**, and the subsequent ones are stored as **m_1**, **m_2** ...

Single condition assignment (Not supported)
-------------------------------------------

    .. code-block:: yaml

        name: assignment ignored
        condition: event.first << event.i == 0
        action:
          debug:
            event: "{{event}}"

| Assignment **cannot** be used for rules that have a single condition, the
| matching event will always be called **event**. In the above example **event.first**
| is ignored and the matching event is stored as **event**. Compare this to multiple

> | condition rules where the matching events are stored as **events**

Reviewer: ```suggestion
| condition rules where the matching events are stored as **events**.
```

| how long does the Rule engine keep the previous event around?

| **Ans:** Currently there is no time limit on how long the rule engine keeps the matched event.
| Once they match they are retracted.

| **Q:** When does the Ansible rulebook stop processing?

| **Ans:** When a Shutdown event is generated from the source plugin, shutdown action is invoked or the process is killed.

| **Q:** Will a condition be evaluated if a variable is missing?

| **Ans:** If a condition refers to an object.attribute which doesn't exist then that condition
| is skipped and not processed.

Example:
    .. code-block:: yaml

        name: send to debug
        condition: event.payload.eventType != 'GET'
        action:
            debug:

In the above case if any of the event.payload.eventType is undefined the condition is
ignored and doesn't match anything.

| **Q:** When a rulebook has multiple rule sets and one of them shuts down are all rule sets terminated?

| **Ans:** Yes, so care should be taken if there are any playbooks running in the other rule sets

| **Q:** How do I check if an attribute in an object referred in a condition exists?

| **Ans:** Use the is defined

> Example:
> 
>     .. code-block:: yaml

Reviewer: ```suggestion
Example:
    .. code-block:: yaml
```

        name: rule1
        condition: event.msg is defined
        action:
          retract_fact:
            fact:
            msg: "{{event.msg}}"

| **Q:** How do I check if an attribute in an object referred in a condition does not exist?

| **Ans:** Use the is not defined

> Example:
> 
>     .. code-block:: yaml

Reviewer: ```suggestion
Example:
    .. code-block:: yaml
```
